Mathlib/GroupTheory/GroupAction/Jordan.lean

```diff
@@ -0,0 +1,641 @@
+    rw [hsa] at hs_prim
+
+    rw [isMultiplyPreprimitive_succ_iff_ofStabilizer G α le_rfl (a := a)]
+    rw [is_one_preprimitive_iff (stabilizer G a) (ofStabilizer G a)]
+    exact IsPreprimitive.of_surjective
+        ofFixingSubgroup_of_singleton_bijective.surjective
+
+  rcases Nat.lt_or_ge (2 * (n + 1)) (Nat.card α) with hn1 | hn2
+
+  · -- CASE where 2 * s.ncard < Nat.card α
+    -- get a, b ∈ s, a ≠ b
+    have : 1 < s.ncard := by rwa [hsn]
+    rw [Set.one_lt_ncard] at this
+    obtain ⟨a, ha, b, hb, hab⟩ := this
+    -- apply rudio to get g ∈ G such that a ∈ g • s, b ∉ g • s
+    obtain ⟨g, hga, hgb⟩ := exists_mem_smul_and_notMem_smul (G := G)
+      s.toFinite hs_nonempty hs_ne_top hab
+
+    let t := s ∩ g • s
+    have ht_prim : IsPreprimitive (fixingSubgroup G t) (ofFixingSubgroup G t) := by
+      apply IsPreprimitive.isPreprimitive_ofFixingSubgroup_inter hs_prim
+      apply Set.ncard_pigeonhole_compl'
+      rwa [Set.ncard_smul_set, hsn, ← two_mul]
+    suffices ∃ m, m < n ∧ t.ncard = m + 1 by
+      obtain ⟨m, hmn, htm⟩ := this
+      apply hrec m hmn hG htm _ ht_prim
+      · apply lt_trans _ hsn'
+        exact Nat.add_lt_add_right hmn 2
+
+    -- from : t ⊆ s, a ∈ t, b ∉ t,
+    -- deduce : 1 ≤ t.ncard < s.ncard
+    use t.ncard - 1
+    suffices t.ncard ≠ 0 by
+      have that : t.ncard - 1 + 1 = t.ncard := Nat.succ_pred_eq_of_ne_zero this
+      simp only [Nat.lt_iff_add_one_le, that, and_true, ge_iff_le]
+      rw [Nat.le_iff_lt_add_one, ← hsn]
+      apply Set.ncard_lt_ncard _ (Set.toFinite s)
+      exact ⟨Set.inter_subset_left, fun h ↦ hgb (Set.inter_subset_right (h hb))⟩
+    · apply Set.ncard_ne_zero_of_mem (a := a)
+      exact ⟨ha, hga⟩
+
+  · -- CASE : 2 * s.ncard ≥ Nat.card α
+    have : Set.Nontrivial sᶜ := by
+      rw [← Set.one_lt_encard_iff_nontrivial, ← sᶜ.toFinite.cast_ncard_eq, Nat.one_lt_cast,
+        ← Nat.add_lt_add_iff_left, Set.ncard_add_ncard_compl, add_comm, hsn, add_comm]
+      exact hsn'
+    -- get a, b ∈ sᶜ, a ≠ b
+    obtain ⟨a, ha : a ∈ sᶜ, b, hb : b ∈ sᶜ, hab⟩ := this
+
+    obtain ⟨g, hga, hgb⟩ := exists_mem_smul_and_notMem_smul (G := G)
+      sᶜ.toFinite (Set.nonempty_of_mem ha)
+      (by intro h
+          simp only [Set.compl_univ_iff] at h
+          simp only [h, Set.not_nonempty_empty] at hs_nonempty)
+      hab
+    let t := s ∩ g • s
+    have : a ∉ s ∪ g • s := by
+      rw [Set.mem_union]
+      intro h
+      rcases h with h | h
+      · exact ha h
+      · rw [Set.smul_set_compl] at hga; exact hga h
+    have ht_prim : IsPreprimitive (fixingSubgroup G t) (ofFixingSubgroup G t) :=
+        IsPreprimitive.isPreprimitive_ofFixingSubgroup_inter hs_prim
+        (by intro h; apply this; rw [h]; trivial)
+
+    suffices ∃ m : ℕ, m < n ∧ t.ncard = m + 1 by
+      obtain ⟨m, hmn, htm⟩ := this
+      exact hrec m hmn hG htm (lt_trans (Nat.add_lt_add_right hmn 2) hsn') ht_prim
+
+    -- from : t ⊆ s, a ∈ t, b ∉ t,
+    -- have : 1 ≤ t.ncard < s.ncard
+    use t.ncard - 1
```

Comment:
... is exactly the same as this part of this proof, so you could extract a common lemma here.